### ISSUE TITLE
[explicit-module-boundary-types] does not respect functions exported within array or object literals

### ISSUE THREAD

**Issue by tornewuff** [typescript-eslint/typescript-eslint#2183]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-module-boundary-types": ["error"]
  }
}
```

```TS
interface Foo {
  f: (x: boolean) => boolean
}

export const a: Foo[] = [
  {
    f: (x) => x,
  },
]
```

**Expected Result**

No errors; the export has an explicit type defined for all members.

**Actual Result**

"Missing return type on function."
"Argument 'x' should be typed."

**Additional Info**

There are no errors when the structure is simpler:

```TS
interface Foo {
  f: (x: boolean) => boolean
}

export const a: Foo = {
  f: (x) => x,
}
```

behaves as expected.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.1.0` |
| `@typescript-eslint/parser`        | `3.1.0` |
| `TypeScript`                       | `3.9.3` |
| `ESLint`                           | `7.1.0` |
| `node`                             | `12.18.0` |
| `npm`                              | `6.14.4` |

**Comment by bradzacher** [comment]
This has been fixed in master, and will be released with the next version.

**Comment by essenmitsosse** [comment]
Just incase someone else stumbles upon this, the same problem occurs in object literals. Here as an example:

```typescript
interface Foo {
    interfaceOuter: {
        interfaceInner: (input: number) => number;
    };
}

export const foo: Foo = {
    interfaceOuter: {
        interfaceInner: (input) => (input),
    },
};
```

**Comment by EtherealData** [comment]
@essenmitsosse 

I am also dealing with this problem inside an object literal. I will have to turn the rule off for the time being
